strax/context.py

```diff
@@ -1844,30 +1844,89 @@ def get_meta(self, run_id, target) -> dict:
 
     get_metadata = get_meta
 
-    def compare_metadata(self, run_id, target, old_metadata):
+    def compare_metadata(self, data1, data2, return_results=False):
         """Compare the metadata between two strax data.
 
-        :param run_id: run id to get
-        :param target: data type to get
-        :param old_metadata: path to metadata to compare, or a dictionary, or a tuple with another
-            run_id, target to compare against the metadata of the first id-target pair
+        :param data1, data2: either a list (tuple) of runid + target pair, or path to metadata to
+        compare,     or a dictionary of the metadata
+        :param return_results: bool, if True, returns a dictionary with metadata and lineages that
+            are found for the inputs
 
         """
 
-        # new metadata for the given runid + target; fetch from context
-        new_metadata = self.get_metadata(run_id, target)
-        # old metadata to compare
-        if isinstance(old_metadata, str):
-            with open(old_metadata) as json_file:
-                old_metadata = json.load(json_file)
-        elif isinstance(old_metadata, dict):
-            old_metadata = old_metadata
-        elif isinstance(old_metadata, (tuple, list)):
-            old_metadata = self.get_metadata(old_metadata[0], old_metadata[1])
-        else:
-            raise ValueError(f"Expected old_metadata as `str` or `dict` got {type(old_metadata)}")
+        def _extract_input(data):
+            """Identify and extract the given input.
+
+            User can either pass a `runid + target` pair or path to metadata json file or a dict
+
+            """
+            if isinstance(data, (tuple, list)):
+                run_id, target = data
+                metafile = None
+            elif isinstance(data, str):
+                run_id, target = None, None
+                metafile = data
+            elif isinstance(data, dict):
+                run_id, target = None, None
+                metafile = data
+            else:
+                raise ValueError(
+                    "data can either be a tuple(list) with runid+target or the path of the metadata"
+                    " json file"
+                )
+            return run_id, target, metafile
+
+        def _extract_metadata_and_lineage(self, run_id, target, metafile):
+            """Extract the actual metadata and lineage based on given inputs and whether the data is
+            available."""
+            # if the runid+target pair is given, check if stored
+            if metafile is None:
+                _is_stored = self.is_stored(run_id, target)
+                metadata = self.get_metadata(run_id, target) if _is_stored else None
```

Comment:
Here you could have used a 

```python
try: 
   self.get_metadata(...)
except strax,DataNotStored:
   ....
```
(or what ever raise we are returning in case it is not stored.) Just as information for the future